In [1]:
from bs4 import BeautifulSoup
import requests
import pandas as pd
import numpy as np
import re

In [4]:
url = 'https://kenpom.com/index.php?y=' + str(2021)
f = requests.get(url)
soup = BeautifulSoup(f.text)

In [3]:
def Kenpom(year):
    
    url = 'https://kenpom.com/index.php?y=' + str(year)
    f = requests.get(url)
    soup = BeautifulSoup(f.text)
    rating_table = soup.find_all('table', {'id': 'ratings-table'})
    thead = rating_table[0].find_all('thead')
    table = rating_table[0]
    for x in thead:
        table = str(table).replace(str(x), '')
    kenpom = pd.read_html(table)[0]
    kenpom.columns = ['Rank', 'Team', 'Conference', 'W-L', 'Adjusted Efficience Margin', 'AdjustOffense', 'AdjustOffense_Rank', 'AdjustDefense', 'AdjustDefense_Rank',
                   'AdjustTempo', 'AdjustTempo_Rank', 'Luck', 'Luck Rank', 'StrengthOfSchedule', 'StrengthOfSchedule_Rank', 'Avg_opponenet_offense', 'Avg_opponenet_offense_rank',
                   'Avg_opponenet_defense', 'Avg_opponenet_defense', 'Non-Conference_SOS', 'Non-Conference_SOS_Rank']
    kenpom = kenpom.set_index('Team')
    return(kenpom)

In [4]:
Kenpom(2015)

Rank Conference   W-L  Adjusted Efficience Margin  \
Team                                                                        
Kentucky 1                 1        SEC  38-1                       36.91   
Wisconsin 1                2        B10  36-4                       33.72   
Duke 1                     3        ACC  35-4                       32.48   
Arizona 2                  4        P12  34-4                       32.36   
Villanova 1                5         BE  33-3                       30.65   
...                      ...        ...   ...                         ...   
San Jose St.             347        MWC  2-28                      -24.59   
Mississippi Valley St.   348       SWAC  6-26                      -25.69   
Central Arkansas         349       Slnd  2-27                      -27.67   
Florida A&M              350       MEAC  2-27                      -31.73   
Grambling St.            351       SWAC  2-27                      -37.65   

                        AdjustOffense  AdjustOffense_Rank  AdjustDefense  \
Team                                                                       
Kentucky 1                      121.3                   6           84.4   
Wisconsin 1                     129.0                   1           95.2   
Duke 1                          124.5                   3           92.0   
Arizona 2                       119.9                   7           87.5   
Villanova 1                     122.7                   4           92.0   
...                               ...                 ...            ...   
San Jose St.                     86.5                 349          111.0   
Mississippi Valley St.           91.6                 337          117.3   
Central Arkansas                 91.0                 341          118.7   
Florida A&M                      85.4                 350          117.1   
Grambling St.                    78.9                 351          116.6   

                        AdjustDefense_Rank  AdjustTempo  AdjustTempo_Rank  \
Team                                                                        
Kentucky 1                               1         62.4               274   
Wisconsin 1                             35         58.7               345   
Duke 1                                  11         65.8               104   
Arizona 2                                3         65.8               102   
Villanova 1                             12         64.2               183   
...                                    ...          ...               ...   
San Jose St.                           302         63.8               201   
Mississippi Valley St.                 346         70.1                 6   
Central Arkansas                       350         69.0                17   
Florida A&M                            345         65.3               122   
Grambling St.                          344         65.4               117   

                         Luck  Luck Rank  StrengthOfSchedule  \
Team                                                           
Kentucky 1              0.048         68                8.58   
Wisconsin 1             0.015        146               11.03   
Duke 1                  0.067         38               11.17   
Arizona 2              -0.033        258                7.37   
Villanova 1             0.031        104                7.67   
...                       ...        ...                 ...   
San Jose St.           -0.030        250                1.45   
Mississippi Valley St. -0.001        187               -7.60   
Central Arkansas        0.004        177               -6.32   
Florida A&M            -0.020        228               -8.50   
Grambling St.          -0.077        327               -8.68   

                        StrengthOfSchedule_Rank  Avg_opponenet_offense  \
Team                                                                     
Kentucky 1                                   26   

In [34]:
kenpom2015 = Kenpom(2015)

In [35]:
kenpom2015.to_csv('KenPomRatings/kenpom2015.csv')

In [36]:
kenpom2016 = Kenpom(2016)

In [37]:
kenpom2016.to_csv('KenPomRatings/kenpom2016.csv')

In [38]:
kenpom2017 = Kenpom(2017)

In [39]:
kenpom2017.to_csv('KenPomRatings/kenpom2017.csv')

In [40]:
kenpom2018 = Kenpom(2018)

In [41]:
kenpom2018.to_csv('KenPomRatings/kenpom2018.csv')

In [42]:
kenpom2019 = Kenpom(2019)

In [43]:
kenpom2019.to_csv('KenPomRatings/kenpom2019.csv')

In [44]:
kenpom2020 = Kenpom(2020)

In [45]:
kenpom2020.to_csv('KenPomRatings/kenpom2020.csv')

In [46]:
kenpom2021 = Kenpom(2021)

In [47]:
kenpom2021.to_csv('KenPomRatings/kenpom2021.csv')

In [48]:
kenpom2022 = Kenpom(2022)

In [49]:
kenpom2022.to_csv('KenPomRatings/kenpom2022.csv')

Basic descriptive statistics on the data set, as appropriate.

- The data is 8 files with Kenpom data from 2015 to 2022 for all 351 Division 1 Ncaa basketball teams. The information on each team is team's kenpom ranking, this is based off of the Adjusted Efficience Margin, the team's name, conferenece, Win-Loss, adjusted efficiency, adjusted offense, adjusted defense, adjusted tempo, luck score, strength of schedule for non-conference and entire schedule as well as ranks for each of the previous categories. The information also includes the oppenents average adjusted offense, defense, and ranks for each. A lot of this information is probably foriegn to someone who isn't a fan of basketball. Basically these are the basic advanced analytics used by basketball experts. 

A brief discussion of the challenges putting together the data set. 

- This data set was difficult to gather because Kenpom doesn't want their data to be scraped. I found this with a lot of sports data. There are tons of ways to get sports data if you are willing to pay a fee. However, without paying for sports data it is difficult to get. Kenpom has much more robust data if you pay for their membership. Kenpom does have an API but I would also have to pay for that.
- Once, I figured out how to decipher the first bits of HTML and get to the section with the dataframe I wanted, it wasn't a difficult scrape. I really enjoy pandas and found a great function called pd.read_html that turned a bunch of random HTML into a nice pandas dataframe. Before finding this command I spent a lot of time putting things in lists and it was taking a lot more time.


Hypotheses that could be tested or explored with this data set. Included specific analyses that could be performed wherever possible. 
- One really cool thing about being able to pull Kenpom data is that you can test future predictions. Kenpom is a dynamic data set. When a game is played the data-set is immediately updated. Without being able to pull the kenpom dataset at a given point, it would be impossible to have a set of games to use the data to predict. When you go back to the Kenpom dataset to input the independent variables the dependent variable you are wanting to test is already included in the dynamic dataset.
- One analysis I would like to do is see if before New Years basketball efficiency stats have any idication on how a team would do in March Madness. I would pull the Kenpom data set before January and wait until March Madness is over in April. Then I could run a regression and see if my model was accurate at all. My perdiction is it is fairly accurate. Last year Baylor won March Madness and they were the number 2 kenpom team going in January 2021.